Fix inpaint only masked pixels are shifted

light-and-ray opened the pull request:
## Description

Fixes bug: pasted image is shifted right and down on factor number of pixels after it was upscaled by this factor

This bug wasn't seen very often because rounded scale factor was almost always 1, and mask blur by default hides this artifact

Also this PR adds option to correct padding to have integer upscale on applying overlay

## Screenshots/videos:
Original:
![original](https://github.com/user-attachments/assets/8d11f6d6-bc15-464b-94a0-9b14910e6d53)

Options: padding=28, mask_blur=0, resolution=451x451, bad quality because I've used LCM, do not care about it

Old behavior:
![old](https://github.com/user-attachments/assets/bcf8253a-0240-475c-92c6-a8e7f5376350)

Integer option on, but bug exists:
![integer_and_NO_fix](https://github.com/user-attachments/assets/b010f3bc-54e1-461c-9ac5-dce461eb0285)
Here scale factor become 2, and it started look like a disaster. And this can happen and happens in current behavior anciently, if scale factor > 1 and upscale non integer 

Integer option on and fixed bug:
![integer_and_fix](https://github.com/user-attachments/assets/8f27302d-2e8c-4e48-bb27-dc143b0f40d9)

Because this integer upscale correction can increase padding on 50-150 pixels, I've made it optional. If option is off, result is still better then the old behavior, but upscale is not integer (worse fit, better resolution):

Integer option off and fixed bug:
![fix](https://github.com/user-attachments/assets/76970b75-dbb0-44ec-8ae6-2323919d2c03)


## Checklist:

- [x] I have read [contributing wiki page](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing)
- [x] I have performed a self-review of my own code
- [x] My code follows the [style guidelines](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Contributing#code-style)
- [x] My code passes [tests](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Tests)

light-and-ray committed:
fix inpaint only masked

w-e-w commented on modules/processing.py:
```diff
@@ -65,8 +65,12 @@ def apply_color_correction(correction, original_image):
-    base_image.paste(image, (x, y))
+    paste_x = max(x - factor_x, 0)
+    paste_y = max(y - factor_y, 0)
+    base_image.paste(image, (paste_x, paste_y))
```
seems to be wrong

I did a inpaing only mask at res 256x256 Denoising 0.01, 0 mask blur, 0 padding
with a mask like this
![image](https://github.com/user-attachments/assets/a0a8db3b-318d-4857-86b3-e0bb239a7616)
WITH this PR the result is visibly misaligned
![20241026-202819-009505 2123871576 0dfb](https://github.com/user-attachments/assets/0cdf57c8-f612-412e-8545-5c879f246788)

whereas without this PR seems to be perfectly aligned
![20241026-202509-740699 4031946088 f9a3](https://github.com/user-attachments/assets/d9f760fa-7296-4b0e-881f-68d8d26354dc)
> if you look closely you could see the parts that are impacted

light-and-ray replied:
Maybe it is, thanks. I'll will to try to figure out

light-and-ray replied:
I could reproduce both... 

When I use mask similar to your + 256p,  cropped region has size (457, 471), generation size is (248, 256) 
When I use mask similar to my (big) + 512p,  cropped region has size (795, 1037), generation size is (448, 584)
(generation aspect ratios are corrected by my extension)

In 1st case my patch make it worse. In 2nd case my patch make it better. And in 1st case uncrop scale factor is even closer to 2: 1.8 vs 1.7

I have no idea WHY. Maybe it's unavoidable in principle, because of resampling, idk. But maybe due to lower resolution is 1st case - more details are gone. And I can add and option for threshold to enable my patch

light-and-ray committed:
add threshold; remove `integer_only_masked`

light-and-ray commented:
I have added a threshold. Because I can't find any other idea why it happens in 256p, except _maybe due to lower resolution is 1st case - more details are gone_. I find it logical

Also I saw this problem that @w-e-w you found in action, after I added aspect ratio correction in my extension which I mentioned above. It sometimes became less then 300-400 pч
\+ I have deleted `integer_only_masked` option in the new commit, because it's not really useful, and buggy.
